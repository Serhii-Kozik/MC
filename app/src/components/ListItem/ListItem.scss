
.table-line {
    width: calc(100% - 32px);
    padding: 8px 16px;
    max-width: 2px 0;
    border-bottom: 1px solid #3498db;
    display: flex;
    justify-content: space-between;

    &.aproaching{

        .col {
            color: #2980b9;
        }
    }

    &.title {
        padding-bottom: 22px;

        .col{
            background: #3498db;
            color: #ffffff;
            transform: skew(20deg);

            .content {
                display: block;
                transform: skew(-20deg);
            }
            
            &.full-name {
                text-align: center;
            }

            &.finish-time {
                justify-content: center;
            }
        }
    }

    &.results {

        .col{

            &.place {
                font-family: Helvetica, sans-serif;
                font-weight: 900;
            }
        }
    }

    .col {
        margin: 0 4px;
        padding: 4px 16px;
        background: #ecf0f1;
        color: #2c3e50;
        font-size: 18px;
        font-family: Verdana, Geneva, Tahoma, sans-serif;

        &.place {
            min-width: 80px;
            text-align: center;
            flex-grow: 1;
            position: relative;

            img {
                position: absolute;
                left: 16px;
                margin-top:3px;
                margin-right: 20px;
                width: 16px;
                height: 16px;
            }
        }

        &.starting-number {
            min-width: 80px;
            text-align: center;
            flex-grow: 1;
        }

        &.full-name {
            min-width: 140px;
            flex-grow: 1;
        }
        &.finish-time {
            display: flex;
            min-width: 200px;
            flex-grow: 1;

            img {
                margin-top:3px;
                margin-right: 20px;
                width: 16px;
                height: 16px;
            }
        }
    }
}
