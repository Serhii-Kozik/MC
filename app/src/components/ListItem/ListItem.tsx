import React from 'react';

import stopwatchIcon from './stopwatch.svg';
import trophyIcon from './trophy-alt.svg';
import './ListItem.scss';

export interface ItemProps {
    title?:             boolean;
    index?:             number;
    place:              number | string;
    athletNumber:       string;
    athletFullName:     string;
    athletFinishTime:   number | string;
}

class ListItem extends React.Component<ItemProps> {

    render(){
        // Set time
        let timeValue: string = '';
        // if recive number - return time string
        if (typeof this.props.athletFinishTime === 'number'){

            const time: Date = new Date(this.props.athletFinishTime);
            timeValue = `${time.getHours()}:${time.getMinutes()}:${time.getSeconds()}:${time.getMilliseconds()}`;

        } else {
            // if recive empty string proceed with it
            timeValue = this.props.athletFinishTime;
        }
        
        // define dynamic CSS classes
        const cssLineClass = this.props.title ? ' table-line title ' : ' table-line results';
        const cssNotFinishedClass = this.props.place === '' ? ' aproaching ' : ' finished ';
        // Render atlet table row
        return (
            <div className={cssLineClass + cssNotFinishedClass}>
                <div className='col place'>
                    <span className='content'>
                        { this.props.title ? '' : <img src={trophyIcon}/>}
                        {this.props.place}
                    </span>
                </div>
                <div className='col starting-number'>
                    <span className='content'>
                        {this.props.athletNumber}
                    </span>
                </div>
                <div className='col full-name'>
                    <span className='content'>
                        {this.props.athletFullName}
                    </span>
                </div>
                <div className='col finish-time'>
                    <span className='content'>
                        {this.props.title ? '' : <img src={stopwatchIcon}/>}
                        {timeValue == null ? 'approaching' : timeValue}
                    </span>
                </div>
            </div>
        );

    }
}

export default ListItem;