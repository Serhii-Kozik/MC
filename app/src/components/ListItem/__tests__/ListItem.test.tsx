import React from 'react';
import ListItem , { ItemProps } from '../ListItem';
import Enzyme ,{ shallow }from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('List component', () => {

    it('should render correctly if title', () => {
        const props: ItemProps =  {
            title:             true,
            index:             1,
            place:             'Place',
            athletNumber:      'F4',
            athletFullName:    'Name',
            athletFinishTime:  'Finish',
          };

        const component = shallow(<ListItem {...props}/>);
        expect(component).toMatchSnapshot();
        expect(component.find('.table-line.title').length).toBe(1);
    });

    it('should render correctly if table row', () => {
        const props: ItemProps =  {
            title:             false,
            index:             1,
            place:             'Place',
            athletNumber:      'F4',
            athletFullName:    'Name',
            athletFinishTime:  '1234343433',
          };

        const component = shallow(<ListItem {...props}/>);
        expect(component.find('.table-line.results').length).toBe(1);
        expect(component.find('.finished').length).toBe(1);

        component.setProps({place:''});
        component.update();
        expect(component.find('.finished').length).toBe(0);
        expect(component.find('.aproaching').length).toBe(1);

    });
});