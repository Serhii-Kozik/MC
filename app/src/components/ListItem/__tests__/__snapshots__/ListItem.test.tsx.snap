// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render correctly if title 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ListItem
    athletFinishTime="Finish"
    athletFullName="Name"
    athletNumber="F4"
    index={1}
    place="Place"
    title={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="col place"
        >
          <span
            className="content"
          >
            
            Place
          </span>
        </div>,
        <div
          className="col starting-number"
        >
          <span
            className="content"
          >
            F4
          </span>
        </div>,
        <div
          className="col full-name"
        >
          <span
            className="content"
          >
            Name
          </span>
        </div>,
        <div
          className="col finish-time"
        >
          <span
            className="content"
          >
            
            Finish
          </span>
        </div>,
      ],
      "className": " table-line title  finished ",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="content"
          >
            
            Place
          </span>,
          "className": "col place",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "",
              "Place",
            ],
            "className": "content",
          },
          "ref": null,
          "rendered": Array [
            "",
            "Place",
          ],
          "type": "span",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="content"
          >
            F4
          </span>,
          "className": "col starting-number",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "F4",
            "className": "content",
          },
          "ref": null,
          "rendered": "F4",
          "type": "span",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="content"
          >
            Name
          </span>,
          "className": "col full-name",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Name",
            "className": "content",
          },
          "ref": null,
          "rendered": "Name",
          "type": "span",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="content"
          >
            
            Finish
          </span>,
          "className": "col finish-time",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "",
              "Finish",
            ],
            "className": "content",
          },
          "ref": null,
          "rendered": Array [
            "",
            "Finish",
          ],
          "type": "span",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="col place"
          >
            <span
              className="content"
            >
              
              Place
            </span>
          </div>,
          <div
            className="col starting-number"
          >
            <span
              className="content"
            >
              F4
            </span>
          </div>,
          <div
            className="col full-name"
          >
            <span
              className="content"
            >
              Name
            </span>
          </div>,
          <div
            className="col finish-time"
          >
            <span
              className="content"
            >
              
              Finish
            </span>
          </div>,
        ],
        "className": " table-line title  finished ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="content"
            >
              
              Place
            </span>,
            "className": "col place",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "",
                "Place",
              ],
              "className": "content",
            },
            "ref": null,
            "rendered": Array [
              "",
              "Place",
            ],
            "type": "span",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="content"
            >
              F4
            </span>,
            "className": "col starting-number",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "F4",
              "className": "content",
            },
            "ref": null,
            "rendered": "F4",
            "type": "span",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="content"
            >
              Name
            </span>,
            "className": "col full-name",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Name",
              "className": "content",
            },
            "ref": null,
            "rendered": "Name",
            "type": "span",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="content"
            >
              
              Finish
            </span>,
            "className": "col finish-time",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "",
                "Finish",
              ],
              "className": "content",
            },
            "ref": null,
            "rendered": Array [
              "",
              "Finish",
            ],
            "type": "span",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
